### PR TITLE
Add macros for BIF boilerplate

### PR DESCRIPTION
**Pull request by blatyo**
I wanted to put this up and get some feedback before going through all the BIFs and updating them. This is also my first time writing a Rust macro, so I'm sure there are things that can be improved.

Addresses part of #294

### CODE REVIEW

File: lumen_runtime/src/otp/maps/put_3.rs [hunk 1]
```diff
@@ -16,57 +16,7 @@ use liblumen_alloc::erts::process::Process;
...
-}
-
-fn module_function_arity() -> Arc<ModuleFunctionArity> {
-    Arc::new(ModuleFunctionArity {
-        module: super::module(),
-        function: function(),
-        arity: 3,
-    })
-}
+bif!(super::module(), "put", native, key, value, map);
```
Comment (blatyo) [inline lumen_runtime/src/otp/maps/put_3.rs]: According to cargo-expand, generates:

``` rust
    fn function() -> Atom {
        Atom::try_from_str("put").unwrap()
    }
    fn module_function_arity() -> Arc<ModuleFunctionArity> {
        Arc::new(ModuleFunctionArity {
            module: super::module(),
            function: function(),
            arity: 1usize + (1usize + (1usize + 0usize)),
        })
    }
    pub(in crate::otp) fn code(arc_process: &Arc<Process>) -> code::Result {
        arc_process.reduce();
        let key = arc_process.stack_pop().unwrap();
        let value = arc_process.stack_pop().unwrap();
        let map = arc_process.stack_pop().unwrap();
        match native(arc_process, key, value, map) {
            Ok(value) => {
                arc_process.return_from_call(value)?;
                Process::call_code(arc_process)
            }
            Err(exception) => result_from_exception(arc_process, exception),
        }
    }
    fn frame() -> Frame {
        Frame::new(module_function_arity(), code)
    }
    pub fn place_frame_with_arguments(
        process: &Process,
        placement: Placement,
        key: Term,
        value: Term,
        map: Term,
    ) -> Result<(), Alloc> {
        process.stack_push(map)?;
        process.stack_push(value)?;
        process.stack_push(key)?;
        process.place_frame(frame(), placement);
        Ok(())
    }
```

File: lumen_runtime/src/otp/maps/from_list_1.rs [hunk 2]
```diff
@@ -15,51 +15,7 @@ use liblumen_alloc::erts::process::Process;
...
-}
-
-fn module_function_arity() -> Arc<ModuleFunctionArity> {
-    Arc::new(ModuleFunctionArity {
-        module: super::module(),
-        function: function(),
-        arity: 1,
-    })
-}
+bif!(super::module(), "from_list", native, list);
```
Comment (blatyo) [inline lumen_runtime/src/otp/maps/from_list_1.rs]: Generates:

``` rust
    fn function() -> Atom {
        Atom::try_from_str("from_list").unwrap()
    }
    fn module_function_arity() -> Arc<ModuleFunctionArity> {
        Arc::new(ModuleFunctionArity {
            module: super::module(),
            function: function(),
            arity: 1usize + 0usize,
        })
    }
    pub(in crate::otp) fn code(arc_process: &Arc<Process>) -> code::Result {
        arc_process.reduce();
        let list = arc_process.stack_pop().unwrap();
        match native(arc_process, list) {
            Ok(value) => {
                arc_process.return_from_call(value)?;
                Process::call_code(arc_process)
            }
            Err(exception) => result_from_exception(arc_process, exception),
        }
    }
    fn frame() -> Frame {
        Frame::new(module_function_arity(), code)
    }
    pub fn place_frame_with_arguments(
        process: &Process,
        placement: Placement,
        list: Term,
    ) -> Result<(), Alloc> {
        process.stack_push(list)?;
        process.place_frame(frame(), placement);
        Ok(())
    }
```

**Comment by KronicDeth** [comment]
I’ve already started work on this, you don’t need to do it.

File: lumen_runtime/src/otp/maps/put_3.rs (hunk 1 above)
Comment (blatyo) [reply lumen_runtime/src/otp/maps/put_3.rs]: My assumption is that `1usize + (1usize + (1usize + 0usize))` will get optimized by the compiler to `3`.

**Comment by KronicDeth** [comment]
https://github.com/lumen/lumen/pull/299 will cover this

**Pull request closed without merging**